frame/balances/src/lib.rs

```diff
@@ -1173,12 +1227,31 @@ impl<T: Config<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I
 	) -> (Self::NegativeImbalance, Self::Balance) {
 		if value.is_zero() { return (NegativeImbalance::zero(), Zero::zero()) }
 
-		Self::mutate_account(who, |account| {
-			// underflow should never happen, but it if does, there's nothing to be done here.
-			let actual = cmp::min(account.reserved, value);
-			account.reserved -= actual;
-			(NegativeImbalance::new(actual), value - actual)
-		})
+		// NOTE: `mutate_account` may fail if it attempts to reduce the balance to the point that an
+		//   account is attempted to be illegally destroyed.
+
+		for attempt in 0..2 {
+			match Self::mutate_account(who, |account| {
+				let best_value = match attempt {
+					0 => value,
+					// If acting as a critical provider (i.e. first attempt failed), then ensure
+					// slash leaves at least the ED.
+					_ => value.min(account.free + account.reserved - T::ExistentialDeposit::get()),
```

Comment:
This is not using `saturating_sub` in contrast to the one further above.
It should be unlikely to end up with free + reserved being less than ED, but let's better be safe than sorry?
```suggestion
					_ => value.min((account.free + account.reserved).saturating_sub(T::ExistentialDeposit::get())),
```

Reply:
The way this can overflow is that the ED is changed in an upgrade, and the accounts are not properly managed. I think it is an important safety to have here.